src/creatures/players/cast/cast_viewer.cpp

```diff
@@ -0,0 +1,1550 @@
+	}
+}
+
+void CastViewer::sendCreatureSay(std::shared_ptr<Creature> creature, SpeakClasses type, const std::string &text, const Position* pos) {
+	if (m_owner) {
+		m_owner->sendCreatureSay(creature, type, text, pos);
+
+		if (type == TALKTYPE_PRIVATE_FROM) {
+			return;
+		}
+
+		for (const auto &it : m_viewers) {
+			it.first->sendCreatureSay(creature, type, text, pos);
+		}
+	}
+}
+
+void CastViewer::disconnectClient(const std::string &message) const {
+	if (m_owner) {
+		m_owner->disconnectClient(message);
+	}
+}
+
+void CastViewer::addViewer(ProtocolGame_ptr client, bool spy) {
+	if (m_viewers.find(client) != m_viewers.end()) {
+		return;
+	}
+
+	std::stringstream message;
+	auto viewerId = m_viewers.size() + 1;
+	message << "Guest-" << viewerId;
+
+	m_viewers[client] = std::make_pair(message.str(), m_id);
+
+	if (!spy) {
+		sendChannelMessage("", message.str() + " has entered the cast.", TALKTYPE_CHANNEL_O, CHANNEL_CAST);
+
+		if (m_viewers.size() > m_castLiveRecord) {
+			m_castLiveRecord = m_viewers.size();
+			sendChannelMessage("", "New record: " + std::to_string(m_castLiveRecord) + " people are watching your livestream now.", TALKTYPE_CHANNEL_O, CHANNEL_CAST);
+		}
+	}
+}
+
+void CastViewer::removeViewer(ProtocolGame_ptr client, bool spy) {
+	auto it = m_viewers.find(client);
+	if (it == m_viewers.end()) {
+		return;
+	}
+
+	auto mit = std::find(m_mutes.begin(), m_mutes.end(), it->second.first);
+	if (mit != m_mutes.end()) {
+		m_mutes.erase(mit);
+	}
+
+	if (!spy) {
+		sendChannelMessage("", it->second.first + " has left the cast.", TALKTYPE_CHANNEL_O, CHANNEL_CAST);
+	}
+
+	m_viewers.erase(it);
+}
+
+void CastViewer::handle(ProtocolGame_ptr client, const std::string &text, uint16_t channelId) {
+	if (!m_owner) {
+		return;
+	}
+
+	auto sit = m_viewers.find(client);
+	if (sit == m_viewers.end()) {
+		return;
+	}
+
+	const int64_t &now = OTSYS_TIME();
+	if (client->m_castCooldownTime + 5000 < now) {
+		client->m_castCooldownTime = now, client->m_castCount = 0;
+	} else if (client->m_castCount++ >= 3) {
+		std::stringstream messageViewer;
+		messageViewer << "Please wait a " << ((client->m_castCooldownTime + 5000 - now) / 1000) + 1 << " seconds to send another message.";
+		client->sendTextMessage(TextMessage(MESSAGE_STATUS, messageViewer.str()));
+		return;
+	}
+
+	bool isCastChannel = channelId == CHANNEL_CAST;
+	if (text[0] == '/') {
+		std::vector<std::string> CommandParam = explodeString(text.substr(1, text.length()), " ", 1);
+		if (strcasecmp(CommandParam[0].c_str(), "show") == 0) {
+			std::stringstream messageViewer;
+			messageViewer << m_viewers.size() << " spectator" << (m_viewers.size() > 1 ? "s" : "") << ". ";
+			for (auto it = m_viewers.begin(); it != m_viewers.end(); ++it) {
+				if (it != m_viewers.begin()) {
+					messageViewer << " ,";
+				}
+
+				messageViewer << it->second.first;
+			}
+
+			messageViewer << ".";
+			client->sendTextMessage(TextMessage(MESSAGE_STATUS, messageViewer.str()));
```

Comment:
I don't know if it is the best way to do this, but with a vector of viewers name we could use `fmt::join` and will be more readable.
```suggestion
			auto viewersNames = std::views::transform(m_viewers, [](const auto &p) {
                                return p.second.first;
                        });
			std::string messageViewer = fmt::format("{} spectator{}, {}.", m_viewers.size(), m_viewers.size() > 1 ? "s" : "", fmt::join(viewersNames.begin(), viewersNames.end(), ", "));

			client->sendTextMessage(TextMessage(MESSAGE_STATUS, messageViewer));
```